src/core/TESTING.md

```diff
@@ -62,8 +62,257 @@ test('my test', async () => {
+
+##### Integration tests
+Depending on the number of external dependencies, you can consider implementing several high-level integration tests. They would work as a set of [smoke tests](https://en.wikipedia.org/wiki/Smoke_testing_(software)) for the most important functionality.
+Subjects for tests:
+- authenticated / unauthenticated access to an endpoint.
+- endpoint validation (params, query, body).
+- main business logic.
+
+###### Functional Test Runner
+If your plugin leverages elasticsearch server to store data and supports additional configuration, you can utilize Functional Test Runner(FTR) to add integration tests. 
+FTR bootstraps the real elasticsearch instance and runs Kibana against it.
+Pros:
+- runs the whole Elastic stack
+- tests cross-plugin integration
+- emulates a real user interaction with the stack
+- allows adjusting config values
+Cons:
+- slow start
+- hard to debug
+- brittle tests
+
+Imagine that `myPlugin` provides a set of endpoints to store & retrieve user data:
+```ts
+router.post(
+  {
+    path: '/myPlugin/something',
+    validate: {
+      body: schema.object({
+        text: schema.string({ maxLength: 100 }),
+      }),
+    },
+  },
+  async (context, request, response) => {
+    const savedObjectsClient = context.core.savedObjects.client;
+    const { id } = await savedObjectsClient.update('myPlugin-type', 'myPlugin', {
+      something: request.body.text,
+    });
+    return response.ok({ body: { id } });
+  }
+);
+
+router.get(
+  {
+    path: '/myPlugin/something/{id}',
+    validate: {
+      params: schema.object({
+        id: schema.string(),
+      }),
+    },
+  },
+  async (context, request, response) => {
+    const savedObjectsClient = context.core.savedObjects.client;
+    const { attributes } = await savedObjectsClient.get('myPlugin-type', request.params.id);
+
+    return response.ok({
+      body: attributes.something,
+    });
+  }
+);
+```
+You can reuse existing [api_integration](/test/api_integration/config.js) setup by registering a test file within a [test loader](/test/api_integration/apis/index.js). More about the existing FTR setup in the [contribution guide](/CONTRIBUTING.md#running-specific-kibana-tests)
+
+The tests cover:
+- authenticated / non-authenticated user access (when applicable)
+```ts
+// TODO after https://github.com/elastic/kibana/pull/53208/
+```
+- request validation
+```ts
+export default function({ getService }: FtrProviderContext) {
+  const supertest = getService('supertest');
+  describe('myPlugin', () => {
+    it('validate params before to store something', async () => {
+      const response = await supertest
+        .post('/myPlugin/something')
+        .set('content-type', 'application/json')
+        .send({ text: 'aaa'.repeat(100) })
+        .expect(400);
+
+      expect(response.body).to.have.property('message');
```

Comment:
NIT: should the exemple test the content of the message, or is it unnecessary?

Reply:
I skipped an assertion against the error message because it can be affected by changes in the validation library. We can add a partial assertion on `longer than 100`. WDYT?

Reply:
sounds good to me